sympy/matrices/expressions/tests/test_matrix_exprs.py

```diff
@@ -242,3 +243,8 @@ def test_Zero_power():
     assert z2**3 == MatPow(z2, 3).doit()
     assert z2**0 == Identity(3)
     raises(ValueError, lambda:MatPow(z2, -1).doit())
+
+
+def test_issue_7421():
+    assert isinstance(diff((D*w)[k,0], w[p,0]), Derivative)
+    assert isinstance(diff(w.T*D*w, w[0,0]), Derivative)
```

Comment:
Based on https://github.com/sympy/sympy/pull/7437#discussion_r12032369, this can't be tested directly because it uses a Dummy. It would be better to test the output directly somehow.

This is a bit worrisome, since it creates a whole class of expressions that can't be compared directly.